### ISSUE TITLE
Feature request: Autodetect target

### ISSUE THREAD

**Issue by palfrey** [google/cargo-raze#14]
Without setting the target setting in Cargo.toml I got the very confusing error "error[E0461]: couldn't find crate `core` with expected target triple x86_64-unknown-linux-gnu" on my Mac. This was confusing, because neither Cargo nor rules_rust required explicit target setting, and so having to work out that I had to explicitly set it for things to not fail horribly was confusing. Also, this now means I've gone from a cross-platform build to a single-platform build, which is somewhat suboptimal.

Thanks for building the tool!

**Comment by acmcarther** [comment]
I thought added a comment already here, but apparently I lost it.

So, you raise two problems. First, **the `target` option is really unclear.** Second, **Raze should support platform-specific dependencies out of the box.** Agreed on both points.

On (1): Do you have an example available? I found the error code [here](https://github.com/rust-lang/rust/blob/dead08cb331343b84564628b139b657f93548320/src/librustc_metadata/diagnostics.rs#L104), but I'm not entirely sure how you ended up with that error. Perhaps you had your own `rust_library` which was inferring the default target from your toolchain (OSX), while your cargo-raze dependencies were defaulting to this linux target? I'm not sure how the OSX toolchain was able to produce a linux rlib in the first place though.

**Comment by palfrey** [comment]
Example: https://github.com/palfrey/potboiler/tree/bazel - the current version of Cargo.toml has no target. In that configuration, all the generated files under "remote" will include the `rustc_flags` "--target=x86_64-unknown-linux-gnu"

**Comment by acmcarther** [comment]
I dont know how that comment got submitted already... Sorry about that. Juggling a couple things at once (I suppose thats status quo for everyone).

**TL;DR: This (2, from my initial comment) is a really hard problem and could really benefit from ideas and use cases to help narrow down the problem space.**

---

For (2), there needs to be a lot more thought. There's a pretty serious mismatch in how Bazel thinks of platform specific dependencies versus how Cargo thinks of them. One solution I had a while ago and haven't have a chance to thoroughly explore is "deferred dependency resolution". This solves the problem by letting individual platforms generate run `cargo-raze` as a repository_rule for their own platform. There are some serious issues with the idea though -- namely that the resolved dependency tree is no longer checked in. This reduces the amount of analysis that tools can do and I personally wouldn't use this feature with that drawback.

A more nuanced approach would be to some how discretize the target space (e.g. reduce it down to [linux, osx, and windows] -- a form that Bazel can understand as [platform_rules](https://docs.bazel.build/versions/master/be/platform.html)), and simply `select(..)` dependencies based upon the applicable platform. This is definitely not a complete victory though -- any other platforms get left out. `Arm` cpu architectures and `musl` toolchains are areas that I'm personally interested in, so this solution doesn't even satisfy me, let alone the wide universe of Rust programmers!

**Comment by photex** [comment]
Hello! Just wanted to chime in here as I decided to experiment with cargo-raze and bazel.

At work we have a monorepo that is inherently a multi-arch build. We essentially only support linux but build for x86_64 and aarch64 (some targets are limited to a single architecture, most will be built for both). We have some bare-metal stuff, but that's currently a separate build.

At the moment the build is cmake which calls cargo as needed.

All this rambling was just to preface the request that anything that helps to clarify how cargo-raze can work with multiple rust target triples would be very helpful. I'm happy to help if I can.

**Comment by UebelAndre** [comment]
Hello, just a friendly ping. Any updates here? Also, are there examples of how `cargo-raze` works with multi-platform projects? Perhaps a workaround could be added to this issue? If not, this feature would be very helpful!

**Comment by UebelAndre** [comment]
While auto detection was never implemented. Raze now supports multi target outputs (as of `0.5.0`). Hopefully this functionality can act as a stopgap until auto detection is implemented
